### PR TITLE
[stdlib] [AutoDiff] More differential operators and utilities

### PR DESCRIPTION
**Pull request by rxwei**
### 1. VJP to `@autodiff` functions

Primitive registration requires function declarations, which are cumbersome for stdlib functions and for closures.  This PR introduces `differentiableFunction(from:)`, which takes a VJP and returns a differentiable `@autodiff` function, which one can use in the middle of any code that is to be differentiated.

The implementation of `differentiableFunction(from:)` is incredibly simple: it's just a wrapper around a pair of primitive registration function declarations.

```swift
@inlinable
public func differentiableFunction<T : Differentiable, R : Differentiable>(
  from vjp: @escaping (T)
           -> (value: R, pullback: (R.CotangentVector) -> T.CotangentVector)
) -> @autodiff (T) -> R {
  @differentiable(vjp: _vjp)
  func original(_ x: T) -> R {
    return vjp(x).value
  }
  func _vjp(_ x: T) -> (R, (R.CotangentVector) -> T.CotangentVector) {
    return vjp(x)
  }
  return original
}
```

It's also super easy to use:

```swift
let diffableFoo: @autodiff (Float) -> Float = differentiableFunction { x in
  return (value: foo(x), pullback: { v in v * x * 2 })
}
```

### 2. Differential operators as methods under `Differentiable`

Top-level differential operators `valueWithPullback(at:in:)`, `pullback(at:in:)`, `valueWithGradient(at:in:)`, and `gradient(at:in:)` have been overloaded for unary up to ternary functions. However, there are cases where the type checker can't infer which differential operator to use even when the trailing closure has a clear arity with concrete types. To ease this tension, we add two sets of differential operators as methods under `Differentiable`:
* `valueWithPullback|pullback|valueWithGradient|gradient(in:)`: This acts as a unary differential operator where the parameter is `self`.
* `valueWithPullback|pullback|valueWithGradient|gradient(at:in:)`: This acts as a binary differential operator where the parameters are `self` and the argument passed in the `at:` position.

### 3. Bug fixes

This patch also fixes the following bugs:
- A crasher in `AdditiveArithmetic` derived conformances. The fix is to check for `hasInterfaceType()` before calling `getType()`.
- A segfault where the autodiff-associated functions are used after free. The fix is to retain associated functions when we fill in an `autodiff_function` instruction in the differentiation transform.

### CODE REVIEW

**Comment by rxwei** [comment]
@jekbradbury you can now use this to make something diffable and get rid of the ugly primitive registrations. :)

File: test/AutoDiff/simple_model.swift [hunk 1]
```diff
@@ -107,7 +107,7 @@ SimpleModelTests.test("gradient") {
   let model = Model(l1: layer, l2: layer, l3: layer)
   let label: Float = 3
   let input: Float = 1
-  let gradModel = gradient(at: model) { (model: Model) -> Float in
```
Comment (dan-zheng) [inline test/AutoDiff/simple_model.swift]: The trailing closure type inference deficiency is unfortunate. 😕

**Review by dan-zheng — APPROVED** [review]
👍 to `differentiableFunction(from:)`! So far, defining custom VJPs has been cumbersome (because we always write `(original(x), { ... })` to call the original function).

The method-style differentiable operators are interesting too!

**Comment by rxwei** [comment]
> So far, defining custom VJPs has been cumbersome (because we always write `(original(x), { ... })` to call the original function).

Yes, this is repetitive and prone to errors when you don't return what the original function is supposed to return. But on the flip side, this is more flexible: if you have a non-differentiable opaque function that you happen know how to implement and want to reuse primal values in the VJP, this allows you to do that.

**Pull request merged**